File: Backup Source/11-12-19/Box.py
'''
Created on 11 Oct 2011

@author: s1152258
'''

from params import *
import utils, wall_algs

import pyximport; pyximport.install()
import numer

class Box():
    def __init__(self, L, lattice_resolution, cell_buffer,
                 D_attract, attract_rate, breakdown_rate,  
                 food_0, food_local_flag, food_conserve_flag,
                 food_pde_flag, D_food, metabolism_rate,
                 density_range, wall_alg):
        self.L = L
        self.cell_buffer = cell_buffer
        self.attract_rate = attract_rate
        self.breakdown_rate = breakdown_rate
        self.food_pde_flag = food_pde_flag
        self.food_local_flag = food_local_flag
        self.food_conserve_flag = food_conserve_flag
        self.metabolism_rate = metabolism_rate
        self.density_range = density_range
        self.wall_alg = wall_alg

        # Dimension
        self.dim = 2

        # Initialise lattice
        self.walls = wall_algs.initialise(lattice_resolution)
        if self.wall_alg == 'blank': wall_algs.blank(self.walls, L)
        elif self.wall_alg == 'funnels': wall_algs.funnels(self.walls, L)
        elif self.wall_alg == 'trap': wall_algs.trap(self.walls, L)
        elif self.wall_alg == 'traps': wall_algs.traps(self.walls, L)
        elif self.wall_alg == 'maze': wall_algs.maze(self.walls, L)

        #     Useful length values
        self.M = self.walls.shape
        self.dx = self.L / self.M[0]

        print('Lattice dx: %f' % self.dx)

        # Initialising chemotaxis
        self.density = np.empty(self.M, dtype=np.float)
        self.density_coeff = 1.0 / (2.0 * self.density_range + 1.0) ** 2.0
        self.field_coeff_arr = np.zeros(self.M, dtype=np.float)
        self._attract_initialise(D_attract)
        self._food_initialise(food_0, D_food)

        # Purely for computational reasons
        self.zeros = np.zeros_like(self.attract)

    def _attract_initialise(self, D):
        self.attract = np.zeros(self.M, dtype=np.float)
        self.attract_coeff_const = D * DELTA_t / self.dx ** 2

    def _food_initialise(self, food_0, D):
        self.food = np.zeros(self.M, dtype=np.float)

        if (self.wall_alg in ['trap', 'traps']) and (self.food_local_flag):
            if self.wall_alg == 'trap':
    #            i_2_4 = self.M[0] // 2

#                i_1_8 = self.M[0] // 8
#                i_3_8 = 3 * i_1_8
#                i_5_8 = 5 * i_1_8
#                self.food[i_3_8 + 1:i_5_8, i_3_8 + 1:i_5_8] = 1.0

#                i_4_8 = 4 * i_1_8
#                self.food[i_4_8, i_4_8] = 1.0

                f_half = 0.5
                f_l = TRAP_LENGTH / L
                i_start = int(round((f_half - f_l / 2.0) * self.walls.shape[0]))
                i_end = int(round((f_half + f_l / 2.0) * self.walls.shape[0]))
                self.food[i_start:i_end + 1, i_start:i_end + 1] = 1 - self.walls[i_start:i_end + 1, i_start:i_end + 1]

#                i_1_10 = self.M[0] // 10
#                i_9_10 = 9 * i_1_10
#                self.food[i_1_10:i_9_10 + 1, i_1_10:i_9_10 + 1] = 1.0

            elif self.wall_alg == 'traps':
#                i_1_5 = self.M[0] // 5
#                i_2_5 = 2 * i_1_5
#                i_3_5 = 3 * i_1_5
#                i_4_5 = 4 * i_1_5            
#                self.food[i_1_5 + 1:i_2_5, i_1_5 + 1:i_2_5] = 1.0
#                self.food[i_3_5 + 1:i_4_5, i_1_5 + 1:i_2_5] = 1.0
#                self.food[i_1_5 + 1:i_2_5, i_3_5 + 1:i_4_5] = 1.0        
#                self.food[i_3_5 + 1:i_4_5, i_3_5 + 1:i_4_5] = 1.0
#
                i_1_10 = self.M[0] // 10
                i_9_10 = 9 * i_1_10
                self.food[i_1_10:i_9_10 + 1, i_1_10:i_9_10 + 1] = 1.0

        else:
            self.food[:, :] = 1.0 - self.walls
        
        if self.food_conserve_flag:
            self.food /= self.food.sum() / self.food.shape[0] ** 2.0

        self.food *= food_0
        
        if self.food_pde_flag:
            self.food_coeff_const = D * DELTA_t / self.dx ** 2

    def arrow_is_find(self, rs):
        return np.asarray((rs + self.L / 2.0) / self.dx, dtype=np.int)

    def arrows_obstructed_find(self, rs):
        i_lattice = self.i_lattice_find(rs)
        i_obstructed = []
        for i_arrow in range(i_lattice.shape[0]):
            try:
                if self.is_wall(i_lattice[i_arrow]):
                    i_obstructed.append(i_arrow)
            except IndexError:
                print('Warning: Invalid lattice index calculated. (%i , %i)' %
                      (i_lattice[i_arrow, 0], i_lattice[i_arrow, 1]))
                i_obstructed.append(i_arrow)
        return i_obstructed, i_lattice

    def is_wall(self, i_cell):
        return self.walls[i_cell[0], i_cell[1]]

    def r_cell_find(self, i_cell):
        return -(self.L / 2.0) + (i_cell + 0.5) * self.dx

    def fields_update(self, rs):
        self._density_update(rs)
        if self.food_pde_flag:
            self._food_update()
        self._attract_update()

    def dx_get(self): return self.dx
    def L_get(self): return self.L
    def cell_buffer_get(self): return self.cell_buffer
    def dim_get(self): return self.dim
    def attract_get(self): return self.attract

#    def _density_update(self, rs):
#        i_lattice = self.i_lattice_find(rs)
#        r = self.density_range
#        self.density[:, :] = 0.0
#        for i_cell in i_lattice:
#            self.density[i_cell[0] - r:i_cell[0] + r + 1, 
#                         i_cell[1] - r:i_cell[1] + r + 1] += self.density_coeff
#        self.density *= 1 - self.walls

    def _density_update(self, rs):
        i_lattice = self.i_lattice_find(rs)
        r = self.density_range
        self.density[:, :] = 0.0
        for i_cell in i_lattice:
            self.density[i_cell[0] - r:i_cell[0] + r + 1, 
                         i_cell[1] - r:i_cell[1] + r + 1] += self.density_coeff
        self.density *= 1 - self.walls

    def _food_update(self):
        if self.food_pde_flag:
            utils.lattice_diffuse(self.walls, self.food, 
                                  self.food_coeff_const, self.field_coeff_arr)
            self.food -= self.metabolism_rate * self.density * DELTA_t
        # Make sure attractant can't be negative
        self.food = np.maximum(self.food, self.zeros)

    def _attract_update(self):
#        utils.lattice_diffuse(self.walls, self.attract, 
#                              self.attract_coeff_const, self.field_coeff_arr)
        numer.lattice_diffuse_cy(self.walls, self.attract, 
                              self.attract_coeff_const, self.field_coeff_arr, self.density_range)
        self.attract += (self.attract_rate * self.density * self.food - 
                         self.breakdown_rate * self.attract) * DELTA_t
        # Make sure attractant can't be negative
        self.attract = np.maximum(self.attract, self.zeros)

    def max_attract_find(self, num_arrows):
        return (self.attract_rate / self.breakdown_rate) * num_arrows * self.food.max()

File: Auxiliary Implementations/C++/src/Box.cpp
/*
 * Box.cpp
 *
 *  Created on: 2 Dec 2011
 *      Author: ejm
 */

#include "Box.hpp"

Box::Box(float L, int lattice_res, float cell_buffer,
				 float D_attract, float attract_rate, float breakdown_rate,
				 float food_0,
				 bool food_pde_flag, float D_food, float metabolism_rate,
				 int density_range) {
	m_L = L;
	m_cell_buffer = cell_buffer;
	m_attract_rate = attract_rate;
	m_breakdown_rate = breakdown_rate;
	m_food_pde_flag = food_pde_flag;
	m_metabolism_rate = metabolism_rate;
	m_density_range = density_range;

	walls_initialise(lattice_res);
	walls_find();

	m_M = m_walls.extent(firstDim);
	m_dx = L / static_cast<float>(m_M);

	m_density.resize(m_walls.shape());
	attract_initialise(D_attract);
	food_initialise(food_0, D_food);
}

bool Box::is_wall(const TinyVector<int, DIM>& i) const {
	return m_walls(i);
}

bool Box::is_wall(const TinyVector<float, DIM>& r) const {
	return m_walls(r_to_i(r));
}

bool Box::is_wall(const Array<float, 1>& r) const {
	return m_walls(r_to_i(r));
}

TinyVector<int, DIM> Box::r_to_i(const TinyVector<float, DIM>& r) const {
	return ((r + m_L / 2.0) / m_dx);
}

TinyVector<int, DIM> Box::r_to_i(const Array<float, 1>& r) const {
	TinyVector<int, DIM> i;
	for (int i_dim = 0; i_dim < DIM; i_dim++)
		i(i_dim) = ((r(i_dim) + m_L / 2.0) / m_dx);
	return i;
}

Array<float, 1> Box::i_to_r(const TinyVector<int, DIM>& i) const {
	Array<float, 1> r(DIM);
	for (int i_dim = 0; i_dim < DIM; i_dim++)
		r(i_dim) = -(m_L / 2.0) + (i(i_dim) + 0.5) * m_dx;
	return r;
}

void Box::fields_update(const Array<float, DIM>& arrow_rs) {
	density_update(arrow_rs);
	attract_update();
	if(m_food_pde_flag)
		food_update();
}

const Array<bool, DIM>& Box::walls_get(void) const {
	return m_walls;
}

const Array<float, DIM>& Box::attract_get(void) const {
	return m_attract;
}

float Box::dx_get(void) const {
	return m_dx;
}

float Box::L_get(void) const {
	return m_L;
}

float Box::cell_buffer_get(void) const {
	return m_cell_buffer;
}

// Private

void Box::walls_initialise(int wall_res) {
	wall_res = 2 * (wall_res / 2) + 1;
	m_walls.resize(wall_res, wall_res);
	m_walls(Range::all(), 0) = true;
	m_walls(Range::all(), wall_res - 1) = true;
	m_walls(wall_res - 1, Range::all()) = true;
	m_walls(0, Range::all()) = true;
}

void Box::walls_find(void) {
    int i_1_8 = m_walls.extent(firstDim) / 8;
    int i_3_8 = 3 * i_1_8;
    int i_4_8 = 4 * i_1_8;
    int i_5_8 = 5 * i_1_8;
    m_walls(Range(i_3_8, i_5_8), i_3_8) = true;
    m_walls(Range(i_3_8, i_5_8), i_5_8) = true;
    m_walls(i_3_8, Range(i_3_8, i_5_8)) = true;
    m_walls(i_5_8, Range(i_3_8, i_5_8)) = true;

    m_walls(i_4_8, i_5_8) = false;
}

void Box::attract_initialise(float D) {
	m_attract.resize(m_walls.shape());
	m_attract_coeff_const = D * DELTA_t / pow(m_dx, 2);
}

void Box::food_initialise(float food_0, float D) {
	m_food.resize(m_walls.shape());
	m_food = 1 - m_walls;
	if (m_food_pde_flag)
		m_food_coeff_const = D * DELTA_t / pow(m_dx, 2);
}

void Box::density_update(const Array<float, DIM>& arrow_rs) {
	m_density = 0.0;
	for (int i_arrow = 0; i_arrow < arrow_rs.extent(secondDim); i_arrow++) {
		TinyVector<int, DIM> i;
		i = r_to_i(arrow_rs(i_arrow, Range::all()));
		for (int i_off_x = -m_density_range; i_off_x <= +m_density_range; i_off_x++) {
			int i_x = i(0) + i_off_x;
			if ((i_x >= 0) and (i_x < m_M))
				for (int i_off_y = -m_density_range; i_off_y <= +m_density_range; i_off_y++) {
					int i_y = i(1) + i_off_y;
					if ((i_y >= 0) and (i_y < m_M) and (not m_walls(i_x, i_y)))
							m_density(i_x, i_y)++;
				}
		}
	}
}

void Box::food_update(void) {
	if(m_food_pde_flag) {
		m_food -= DELTA_t * m_metabolism_rate * m_density;
		lattice_diffuse(m_food, m_food_coeff_const);
		for(int i_x = 0; i_x < m_M; i_x++)
			for(int i_y = 0; i_y < m_M; i_y++)
				if(m_food(i_x, i_y) < 0.0)
					m_food(i_x, i_y) = 0.0;
	}
}

void Box::attract_update(void) {
	lattice_diffuse(m_attract, m_attract_coeff_const);
	m_attract += DELTA_t * (
			m_attract_rate * m_density * m_food -
			m_breakdown_rate * m_attract);
	for(int i_x = 0; i_x < m_M; i_x++)
		for(int i_y = 0; i_y < m_M; i_y++)
			if (m_attract(i_x, i_y) < 0.0) m_attract(i_x, i_y) = 0.0;
}

void Box::lattice_diffuse(Array<float, DIM>& field, float coeff) {
	Array<float, DIM> field_coeff_arr(field.shape());
	field_coeff_arr = 0.0;
	for (int i_x = 1; i_x < m_M - 1; i_x++)
		for (int i_y = 1; i_y < m_M - 1; i_y++) {
			float field_cur = field(i_x, i_y);
			field_coeff_arr(i_x, i_y) = (not m_walls(i_x, i_y)) *
				((not m_walls(i_x + 1, i_y)) * (field(i_x + 1, i_y) - field_cur) +
				 (not m_walls(i_x - 1, i_y)) * (field(i_x - 1, i_y) - field_cur) +
				 (not m_walls(i_x, i_y + 1)) * (field(i_x, i_y + 1) - field_cur) +
				 (not m_walls(i_x, i_y - 1)) * (field(i_x + 1, i_y) - field_cur));
		}
	field += coeff * field_coeff_arr;
}
